sync-core/src/main/java/com/cloudant/sync/query/IndexUpdater.java

```diff
@@ -248,53 +249,52 @@ public Boolean call() {
         }
 
         List<DBParameter> parameters = new ArrayList<DBParameter>();
-        if (arrayCount == 0) {
-            // The are no arrays in the values we are indexing. We just need to index the fields
-            // in the index. _id and _rev are special fields in that they don't appear in the
-            // body, so they need special-casing to get the values.
-
-            List<String> initialIncludedFields = new ArrayList<String>();
-            initialIncludedFields.add("_id");
-            initialIncludedFields.add("_rev");
-            List<Object> initialArgs = new ArrayList<Object>();
-            initialArgs.add(rev.getId());
-            initialArgs.add(rev.getRevision());
-            DBParameter parameter = populateDBParameter(fieldNames,
-                                                        initialIncludedFields,
-                                                        initialArgs,
-                                                        indexName,
-                                                        rev);
-            if (parameter == null) {
-                return null;
-            }
-            parameters.add(parameter);
-        } else if (arrayFieldName != null) {
-            // We know the value is an array, we found this out in the check above
+
+        if (arrayCount == 1) {
+            // Value is an array
             List<Object> arrayFieldValues;
-            arrayFieldValues = (ArrayList) ValueExtractor.extractValueForFieldName(arrayFieldName,
-                                                                                   rev.getBody());
-            for (Object value: arrayFieldValues) {
-                List<String> initialIncludedFields = new ArrayList<String>();
-                initialIncludedFields.add("_id");
-                initialIncludedFields.add("_rev");
-                initialIncludedFields.add(arrayFieldName);
-                List<Object> initialArgs = new ArrayList<Object>();
-                initialArgs.add(rev.getId());
-                initialArgs.add(rev.getRevision());
-                initialArgs.add(value);
-                DBParameter parameter;
-                parameter = populateDBParameter(fieldNames,
-                                                initialIncludedFields,
-                                                initialArgs,
-                                                indexName,
-                                                rev);
-                if (parameter == null) {
-                    return null;
+            arrayFieldValues = (List) ValueExtractor.extractValueForFieldName(arrayFieldName,
+                                                                              rev.getBody());
+            if (arrayFieldValues != null && arrayFieldValues.size() > 0) {
```

Comment:
Actually this line of code is what, in part, fixes our empty array issue.  The check looks to see if there are values in arrayFieldValues.  If there are, then array processing happens as before.  But in the event that the array is empty, processing falls through to the field processing section which then handles an empty array as it would a missing field value.  The check for null followed by the zero size check is standard in Java since it is bad practice to check a size of 0 before knowing you have a non-Null object.  It is essentially the same logic as in CDTDatastore https://github.com/cloudant/CDTDatastore/blob/48139-query-fix-empty-array-field-indexing/Classes/common/query/CDTQIndexUpdater.m#L339.